### PR TITLE
 Report all the js file errors and skip only the nodes that are not allowed in JS file

### LINKED ISSUE #11800: JavaScript: types allowed when decorator is added to a JavaScript file
**TypeScript Version:** nightly (2.1.0-dev.201xxxxx)

When I add a decorator to a .js file, I no longer get the `'types' can only be used in a .ts file` error. Is this expected behavior?

**Code**

``` js

@SomeDecorator
class SomeClass {
    foo(x:number) { 

    }
}

```

**Expected behavior:**
`'types' can only be used in a .ts file`

**Actual behavior:** 
no errors

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #11800

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -773,48 +773,89 @@ namespace ts {
...
                 return diagnostics;
 
-                function walk(node: Node): boolean {
-                    if (!node) {
-                        return false;
+                function walk(node: Node) {
+                    // Return directly from the case if the given node doesnt want to visit each child
+                    // Otherwise break to visit each child
+
+                    switch (parent.kind) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: why are we using parent instead of node?

Comment (mhegazy) [inline src/compiler/program.ts]: but what was wrong with the previous pattern where we check parameter node, and then check it has a questiontoken set?

Comment (sheetalkamat) [inline src/compiler/program.ts]: There was no control over what to skip.. With this we visit all nodes except the nodes that aren't supported. (Eg. checkout how we use to return true when we found ? token, which would mean we wouldn't look into parameter children.. While that might be ok there it isn't if say for example we want to report error say on decorator of the class declaration and also verify say methods of class..) With this if we have say found typeNode of the variable declaration/function return type we don't go it that type node. But still visit other nodes of the variable declaration/function declaration which means we would be reporting other errors.

File: src/compiler/program.ts [hunk 2]
```diff
@@ -824,124 +865,82 @@ namespace ts {
...
+                            case SyntaxKind.DeclareKeyword:
+                                diagnostics.push(createDiagnosticForNode(modifier, Diagnostics._0_can_only_be_used_in_a_ts_file, tokenToString(modifier.kind)));
+                                break;
+
+                            // These are all legal modifiers.
+                            case SyntaxKind.StaticKeyword:
+                            case SyntaxKind.ExportKeyword:
+                            case SyntaxKind.ConstKeyword:
+                            case SyntaxKind.DefaultKeyword:
+                            case SyntaxKind.AbstractKeyword:
```
Comment (mhegazy) [inline src/compiler/program.ts]: why is `abstract` and `const` legal modifiers?

Comment (sheetalkamat) [inline src/compiler/program.ts]: hmm, I just copied from earlier implementation.. ```const``` should be valid when variable statement to handle ```const a = 10``` and ```abstract``` shouldn't be valid. I will update this

File: src/compiler/program.ts [hunk 3]
```diff
@@ -870,11 +870,12 @@ namespace ts {
...
                                 return;
                             }
-                        // pass through
+                            // pass through
+                            let isConstValid = false;
                         case SyntaxKind.VariableStatement:
                             // Check modifiers
                             if (nodes === (<ClassDeclaration | FunctionLikeDeclaration | VariableStatement>parent).modifiers) {
-                                return checkModifiers(<NodeArray<Modifier>>nodes);
+                                return checkModifiers(<NodeArray<Modifier>>nodes, isConstValid);
```
Comment (mhegazy) [inline src/compiler/program.ts]: is not this always false?

**Comment by sheetalkamat** [comment]
@dotnet-bot test this please

File: src/compiler/program.ts (hunk 3 above)
Comment (sheetalkamat) [inline src/compiler/program.ts]: forgot to amend the changes before pushing it out. Updated it.

File: src/compiler/program.ts [hunk 4]
```diff
@@ -959,124 +1000,87 @@ namespace ts {
...
-                            if (parameter.questionToken) {
-                                diagnostics.push(createDiagnosticForNode(parameter.questionToken, Diagnostics._0_can_only_be_used_in_a_ts_file, "?"));
-                                return true;
-                            }
-                            if (parameter.type) {
-                                diagnostics.push(createDiagnosticForNode(parameter.type, Diagnostics.types_can_only_be_used_in_a_ts_file));
-                                return true;
+                            // Check modifiers
+                            if (nodes === (<ClassDeclaration | FunctionLikeDeclaration | VariableStatement>parent).modifiers) {
+                                return checkModifiers(<NodeArray<Modifier>>nodes, !isConstInvalid);
```
Comment (mhegazy) [inline src/compiler/program.ts]:
```ts
return checkModifiers(<NodeArray<Modifier>>nodes, parent.kind === SyntaxKind.VariableStatement)
```

Comment (mhegazy) [inline src/compiler/program.ts]: it much more readable and better than accessing an undefined variable to be frank.

**Pull request merged**